File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ShippingInputTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ShippingInputTyped {
    public Dollar totalPrice;
    public Dollar shippingCost;
    public String notes;

    public ShippingInputTyped(Dollar totalPrice, Dollar shippingCost, String notes) {
        this.totalPrice = totalPrice;
        this.shippingCost = shippingCost;
        this.notes = notes;
    }

    public ShippingInputTyped(ShippingInputString s) {
        this.totalPrice = new Dollar(s.totalPrice);
        this.shippingCost = new Dollar(s.shippingCost);
        this.notes = s.notes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ShippingInputTyped)) return false;
        ShippingInputTyped that = (ShippingInputTyped) o;
        return Objects.equals(totalPrice, that.totalPrice)
            && Objects.equals(shippingCost, that.shippingCost)
            && Objects.equals(notes, that.notes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(totalPrice, shippingCost, notes);
    }

    public ShippingInputString toShippingInputString() {
        return new ShippingInputString(String.valueOf(totalPrice), String.valueOf(shippingCost), String.valueOf(notes));
    }

    public static List<ShippingInputTyped> fromStringList(List<ShippingInputString> list) {
        List<ShippingInputTyped> result = new ArrayList<>();
        for (ShippingInputString s : list) result.add(new ShippingInputTyped(s));
        return result;
    }

    public static List<ShippingInputString> toStringList(List<ShippingInputTyped> list) {
        List<ShippingInputString> result = new ArrayList<>();
        for (ShippingInputTyped t : list) result.add(t.toShippingInputString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("totalPrice", totalPrice == null ? null : totalPrice.value);
        m.put("shippingCost", shippingCost == null ? null : shippingCost.value);
        m.put("notes", notes);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static ShippingInputTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new ShippingInputTyped(
            new Dollar(Json.asString(Json.require(m, "totalPrice"), "totalPrice")),
            new Dollar(Json.asString(Json.require(m, "shippingCost"), "shippingCost")),
            Json.asString(Json.require(m, "notes"), "notes"));
    }

    public static ShippingInputTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<ShippingInputTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (ShippingInputTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<ShippingInputTyped> fromJSONList(String json) {
        List<ShippingInputTyped> result = new ArrayList<ShippingInputTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "ShippingInputTyped")));
        return result;
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/ShippingInputTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class ShippingInputTyped
    {
        public Dollar totalPrice;
        public Dollar shippingCost;
        public string notes;

        public ShippingInputTyped(Dollar totalPrice, Dollar shippingCost, string notes)
        {
            this.totalPrice = totalPrice;
            this.shippingCost = shippingCost;
            this.notes = notes;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("totalPrice", Json.ToText(this.totalPrice));
            w.WriteString("shippingCost", Json.ToText(this.shippingCost));
            w.WriteString("notes", this.notes);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static ShippingInputTyped FromJsonElement(JsonElement m)
        {
            return new ShippingInputTyped(
                new Dollar(Json.AsString(Json.Require(m, "totalPrice"), "totalPrice")),
                new Dollar(Json.AsString(Json.Require(m, "shippingCost"), "shippingCost")),
                Json.AsString(Json.Require(m, "notes"), "notes")
            );
        }

        public static ShippingInputTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<ShippingInputTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<ShippingInputTyped> FromJSONList(string json)
        {
            var result = new List<ShippingInputTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "ShippingInputTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Total Price={totalPrice}, Shipping Cost={shippingCost}, Notes={notes}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not ShippingInputTyped other) return false;
            return object.Equals(this.totalPrice, other.totalPrice)
                && object.Equals(this.shippingCost, other.shippingCost)
                && object.Equals(this.notes, other.notes);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.totalPrice);
            h.Add(this.shippingCost);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}
